src/visualization.c

```diff
@@ -15,25 +15,46 @@
 // You should have received a copy of the GNU General Public License along
 // with visdriver. If not, see <https://www.gnu.org/licenses/>.
 
+#if defined(_MSC_VER)
+#define _USE_MATH_DEFINES // for M_PI from math.h
+#else
+#define _GNU_SOURCE // for M_PI from math.h
+#endif
+
+#include <math.h>
 #include <stdint.h>
 #include <string.h> // memset
 
+#include <kissfft/kiss_fftr.h>
+
 #include "log.h"
 #include "visualization.h"
 
 #define VIS_FRAMES 576 // dictated by vis.h
 
 winampVisModule *g_active_vis_module = NULL;
+static kiss_fftr_cfg g_kiss_fft_cfg = NULL;
+static int16_t g_prev_interleaved[VIS_FRAMES * 2];
 
 void __cdecl SAVSAInit(int maxlatency_in_ms, int srate) {
   log_debug("Input plugin announced: Maximum latency %dms, sampling rate %d "
             "(SAVSAInit).",
             maxlatency_in_ms, srate);
   g_active_vis_module->sRate = srate;
+  memset(g_prev_interleaved, 0, sizeof(g_prev_interleaved));
+
+  g_kiss_fft_cfg = kiss_fftr_alloc(VIS_FRAMES * 2, 0, NULL, NULL);
 }
 
 void __cdecl SAVSADeInit() {
-  // Nothing to do
+  kiss_fftr_free(g_kiss_fft_cfg);
+  g_kiss_fft_cfg = NULL;
+}
+
+static kiss_fft_scalar hann_factor(size_t index, size_t samples) {
```

Comment:
@hartwork It is a relatively "heavy" computation here done for each sample of each frame. I ough you to precompute or cache the values to gain performances.

Reply:
@BenPyton that's a nice idea! I have made pull request #15 now. I think I may try to benchmark the two because part of me thinks that access to memory is also expensive and it would be great to be sure. In case you do benchmarking before me, please let me know, would definitely be interesting.

Reply:
@hartwork It's right, you are using a function to return the value, so it may not be beneficial to replace by just an array value returned.
In OsciView, I use directly the precomputed array in the loop, thus the processor can cache it for better access speed.

Reply:
@BenPyton I did some benchmarking now, and the pre-computed version seems to be about 7x to 9x faster, *if* my benchmark is alright.

Here is what I see:

```console
# gcc -std=c99 -Wall -Wextra -pedantic -lm main.c && for i in live cached ; do time ./a.out 1000000 "${i}"; done
Live version, 1000000 passes...
Done, checksum 653228.

real    0m38.047s
user    0m38.029s
sys     0m0.009s
Cached version, 1000000 passes...
Done, checksum 653228.

real    0m4.135s
user    0m4.131s
sys     0m0.004s
```

Here's the C code for `main.c`:

```c
#define _GNU_SOURCE // for M_PI from math.h

#define VIS_SAMPLES 756

#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h> // atoi
#include <string.h>

static float hann_factor(size_t index, size_t samples) {
  return 1 / 2.0f *
         (1 + cosf(2.0f * (float)M_PI * (index - samples / 2) / samples));
}

static void usage() { fprintf(stderr, "usage: PASSES {cached,live}\n"); }

int main(int argc, char **argv) {
  if (argc != 3) {
    usage();
    return 1;
  }

  int16_t source[VIS_SAMPLES * 2];

  for (int sample = 0; sample < VIS_SAMPLES * 2; sample++) {
    source[sample] = sample; // some deterministic dummy
  }

  float target[VIS_SAMPLES * 2];

  const int passes = atoi(argv[1]);
  if (strcmp(argv[2], "cached") == 0) {
    printf("Cached version, %d passes...\n", passes);
    float hann_window[VIS_SAMPLES * 2];

    for (int sample = 0; sample < VIS_SAMPLES * 2; sample++) {
      hann_window[sample] = hann_factor(sample, VIS_SAMPLES * 2);
    }

    for (int pass = 0; pass < passes; pass++) {
      for (int sample = 0; sample < VIS_SAMPLES * 2; sample++) {
        target[sample] = source[sample] * hann_window[sample];
      }
    }
  } else if (strcmp(argv[2], "live") == 0) {
    printf("Live version, %d passes...\n", passes);
    for (int pass = 0; pass < passes; pass++) {
      for (int sample = 0; sample < VIS_SAMPLES * 2; sample++) {
        target[sample] = source[sample] * hann_factor(sample, VIS_SAMPLES * 2);
      }
    }
  } else {
    usage();
    return 1;
  }

  // Make all of target read once
  unsigned int checksum = 0;
  for (int sample = 0; sample < VIS_SAMPLES * 2; sample++) {
    checksum += target[sample];
  }

  printf("Done, checksum %d.\n", checksum);
  return 0;
}
```

What do you think?

Reply:
@hartwork I think it is no doubt to use cached version.
The only thing I can say about your benchmark is that you are also counting the pre-computation of the hanning for the cached version. Since it is done once and does not impact perfs for each frames, you should not count it in the benchmark imho.
However, even with this extra time, it is largely faster than no cached hanning.

Reply:
@BenPyton thanks for the review! I'll go merge #15 now. Thanks for the tip on pre-computation! :+1:

Reply:
@hartwork 
I have more time now to explain more about your benchmarking time issue.
You are comparing the total time taken by the whole benchmark program to complete, and not only the thing you want to compare (the for loop applying the hanning).
Here a quick diagram I made to show you : 
![benchmark_profiling](https://user-images.githubusercontent.com/13402355/220887852-bf3a5a7c-2d35-457e-b477-71925916b8f8.png)
(it is exagerated here, the orange part is much bigger than that)

On this diagram, the orange part is what we want to compare (what is really done each frame).
So if you really want to do quantitative comparison, you should get precise time just before and after (on the red marks) to get the real work duration.
For that you can use for example [`clock_gettime()`](https://www.gnu.org/software/libc/manual/html_node/Getting-the-Time.html#index-clock_005fgettime) to get the time before and after, and then subtract them to get the elapsed time, for example like this (didn't test the code): 
```c
#include <time.h>
struct timespec before, after;
clock_gettime(CLOCK_REALTIME, &before);
// what we want to measure
clock_gettime(CLOCK_REALTIME, &after);
double elasped_time = (after.tv_sec - before.tv_sec) // whole seconds
    + (double)(after.tv_nsec - before.tv_nsec) / 1000000000.0; // sub-second part
```
and then you `printf` the result.

However, what you did is still a good qualitative comparison 😉

Reply:
@BenPyton yes, I agree, in some sense my version was a shortcut. The 1,000,000 iterations reduce that effect fortunately.

Reply:
Yes indeed! For video games I am used to optimize very small things where a simple `printf` takes usually much more time 😅 
(And generally in games we need to have the full game to run to try some optimizations)
But yeah, your benchmark (though it has not the exact time gain values) is qualitatively good! There is no doubt with it 😄